reference/test/preconditioner/ilu.cpp

```diff
@@ -258,145 +264,201 @@ TEST_F(Ilu, SolvesCustomTypeDefaultFactorySingleRhs)
 
     preconditioner->apply(alpha_linop.get(), b.get(), beta_linop.get(),
                           x.get());
 
-    GKO_ASSERT_MTX_NEAR(x.get(), l({-7.75, 6.25, 1.5}), 1e-14);
+    GKO_ASSERT_MTX_NEAR(x.get(), l({-7.75, 6.25, 1.5}),
+                        r<TypeParam>::value * 1e+1);
 }
 
 
-TEST_F(Ilu, SolvesMultipleRhs)
+TYPED_TEST(Ilu, SolvesMultipleRhs)
 {
-    const auto b =
-        gko::initialize<Mtx>({{1.0, 8.0}, {3.0, 21.0}, {6.0, 24.0}}, exec);
-    auto x = Mtx::create(exec, gko::dim<2>{3, 2});
+    using Mtx = typename TestFixture::Mtx;
+    using T = typename TestFixture::value_type;
+    const auto b = gko::initialize<Mtx>(
+        {I<T>{1.0, 8.0}, I<T>{3.0, 21.0}, I<T>{6.0, 24.0}}, this->exec);
+    auto x = Mtx::create(this->exec, gko::dim<2>{3, 2});
     x->copy_from(b.get());
-    auto preconditioner = ilu_pre_factory->generate(l_u_composition);
+    auto preconditioner =
+        this->ilu_pre_factory->generate(this->l_u_composition);
 
     preconditioner->apply(b.get(), x.get());
 
     GKO_ASSERT_MTX_NEAR(x.get(), l({{-0.125, 2.0}, {0.25, 3.0}, {1.0, 1.0}}),
-                        1e-14);
+                        r<TypeParam>::value * 1e+1);
 }
 
 
-TEST_F(Ilu, SolvesDifferentNumberOfRhs)
+TYPED_TEST(Ilu, SolvesDifferentNumberOfRhs)
 {
-    const auto b1 = gko::initialize<Mtx>({-3.0, 6.0, 9.0}, exec);
-    auto x11 = Mtx::create(exec, gko::dim<2>{3, 1});
-    auto x12 = Mtx::create(exec, gko::dim<2>{3, 1});
+    using Mtx = typename TestFixture::Mtx;
+    using T = typename TestFixture::value_type;
+    const auto b1 = gko::initialize<Mtx>({-3.0, 6.0, 9.0}, this->exec);
+    auto x11 = Mtx::create(this->exec, gko::dim<2>{3, 1});
+    auto x12 = Mtx::create(this->exec, gko::dim<2>{3, 1});
     x11->copy_from(b1.get());
     x12->copy_from(b1.get());
-    const auto b2 =
-        gko::initialize<Mtx>({{1.0, 8.0}, {3.0, 21.0}, {6.0, 24.0}}, exec);
-    auto x2 = Mtx::create(exec, gko::dim<2>{3, 2});
+    const auto b2 = gko::initialize<Mtx>(
+        {I<T>{1.0, 8.0}, I<T>{3.0, 21.0}, I<T>{6.0, 24.0}}, this->exec);
+    auto x2 = Mtx::create(this->exec, gko::dim<2>{3, 2});
     x2->copy_from(b2.get());
-    auto preconditioner = ilu_pre_factory->generate(l_u_composition);
+    auto preconditioner =
+        this->ilu_pre_factory->generate(this->l_u_composition);
 
     preconditioner->apply(b1.get(), x11.get());
     preconditioner->apply(b2.get(), x2.get());
     preconditioner->apply(b1.get(), x12.get());
 
-    GKO_ASSERT_MTX_NEAR(x11.get(), l({-3.0, 2.0, 1.0}), 1e-14);
+    GKO_ASSERT_MTX_NEAR(x11.get(), l({-3.0, 2.0, 1.0}),
+                        r<TypeParam>::value * 1e+1);
     GKO_ASSERT_MTX_NEAR(x2.get(), l({{-0.125, 2.0}, {0.25, 3.0}, {1.0, 1.0}}),
-                        1e-14);
-    GKO_ASSERT_MTX_NEAR(x12.get(), x11.get(), 1e-14);
+                        r<TypeParam>::value * 1e+1);
+    GKO_ASSERT_MTX_NEAR(x12.get(), x11.get(), r<TypeParam>::value * 1e+1);
```

Comment:
should `r<TypeParam>::value * 1e+1` be `r<TypeParam>::value` to keep `1e-14` in double?

Reply:
This fails in single if I do that. Single precision seems to be closer to `1e-6`.

Reply:
got it. thanks